File: dotnet/src/SemanticKernel/AI/Embeddings/IEmbeddingGenerator.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.Diagnostics;

namespace Microsoft.SemanticKernel.AI.Embeddings;

/// <summary>
/// Represents a generator of embeddings.
/// </summary>
/// <typeparam name="TValue">The type from which embeddings will be generated.</typeparam>
/// <typeparam name="TEmbedding">The numeric type of the embedding data.</typeparam>
public interface IEmbeddingGenerator<TValue, TEmbedding>
    where TEmbedding : unmanaged
{
    /// <summary>
    /// Generates an embedding from the given <paramref name="data"/>.
    /// </summary>
    /// <param name="data">List of strings to generate embeddings for</param>
    /// <returns>List of embeddings</returns>
    Task<IList<Embedding<TEmbedding>>> GenerateEmbeddingsAsync(IList<TValue> data);
}

/// <summary>
/// Provides a collection of static methods for operating on <see cref="IEmbeddingGenerator{TValue, TEmbedding}"/> objects.
/// </summary>
public static class EmbeddingGeneratorExtensions
{
    /// <summary>
    /// Generates an embedding from the given <paramref name="value"/>.
    /// </summary>
    /// <typeparam name="TValue">The type from which embeddings will be generated.</typeparam>
    /// <typeparam name="TEmbedding">The numeric type of the embedding data.</typeparam>
    /// <param name="generator">The embedding generator.</param>
    /// <param name="value">A value from which an <see cref="Embedding{TEmbedding}"/> will be generated.</param>
    /// <returns>A list of <see cref="Embedding{TEmbedding}"/> structs representing the input <paramref name="value"/>.</returns>
    public static async Task<Embedding<TEmbedding>> GenerateEmbeddingAsync<TValue, TEmbedding>
        (this IEmbeddingGenerator<TValue, TEmbedding> generator, TValue value)
        where TEmbedding : unmanaged
    {
        Verify.NotNull(generator, "Embeddings generator cannot be NULL");
        return (await generator.GenerateEmbeddingsAsync(new[] { value })).FirstOrDefault();
    }
}

File: typescript/src/ai/embeddings/iEmbeddingGenerator.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { Embedding } from './embedding';

/**
 * Represents a generator of embeddings.
 * @template TValue The type from which embeddings will be generated.
 */
interface IEmbeddingGenerator<TValue = string> {
    /**
     * Generates an embedding from the given data.
     * @param data List of strings to generate embeddings for
     * @returns List of embeddings
     */
    generateEmbeddings(data: TValue[]): Promise<Embedding[]>;
}

/**
 * Provides a collection of static methods for operating on IEmbeddingGenerator objects.
 */
export class EmbeddingGeneratorExtensions {
    /**
     * Generates an embedding from the given value.
     * @template TValue The type from which embeddings will be generated.
     * @param generator The embedding generator.
     * @param value A value from which an Embedding will be generated.
     * @returns A list of Embedding structs representing the input value.
     */
    static GenerateEmbeddingAsync<TValue = string>(
        generator: IEmbeddingGenerator<TValue>,
        value: TValue
    ): Promise<Embedding[]> {
        if (!generator) {
            throw new Error('Embeddings generator cannot be NULL');
        }
        return generator.generateEmbeddings([value]);
    }
}
